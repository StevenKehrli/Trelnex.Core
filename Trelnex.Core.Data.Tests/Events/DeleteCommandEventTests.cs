using Snapshooter.NUnit;

namespace Trelnex.Core.Data.Tests.Events;

public class DeleteCommandEventTests
{
    private readonly string _typeName = "test-item";

    [Test]
    public async Task DeleteCommandEvent()
    {
        var id = "bc4971e4-6dae-45ba-b3ee-43c036e0d957";
        var partitionKey = "60525a8e-b084-4a37-b461-d08330760ef2";

        var startDateTime = DateTime.UtcNow;

        var requestContext = TestRequestContext.Create();

        // create our command provider
        var factory = await InMemoryCommandProviderFactory.Create();

        var commandProvider = factory.Create<ITestItem, TestItem>(
                typeName: _typeName,
                commandOperations: CommandOperations.Delete);

        var createCommand = commandProvider.Create(
            id: id,
            partitionKey: partitionKey);

        createCommand.Item.PublicId = 1;
        createCommand.Item.PublicMessage = "Public #1";
        createCommand.Item.PrivateMessage = "Private #1";

        // save it
        await createCommand.SaveAsync(
            requestContext: requestContext,
            cancellationToken: default);

        var deleteCommand = await commandProvider.DeleteAsync(
            id: id,
            partitionKey: partitionKey);

        Assert.That(deleteCommand, Is.Not.Null);
        Assert.That(deleteCommand!.Item, Is.Not.Null);

        // save it
        await deleteCommand.SaveAsync(
            requestContext: requestContext,
            cancellationToken: default);

        // get the events
        var events = (commandProvider as InMemoryCommandProvider<ITestItem, TestItem>)!.GetEvents();

        // snapshooter does a poor job of the serialization of dynamic
        // so explicit check of the changes array

        Assert.Multiple(() =>
        {
            Assert.That(
                events[0].Changes!,
                Has.Length.EqualTo(2));

            Assert.That(
                events[0].Changes![0].OldValue!.GetInt32(),
                Is.EqualTo(0));

            Assert.That(
                events[0].Changes![0].NewValue!.GetInt32(),
                Is.EqualTo(1));

            Assert.That(
                events[0].Changes![1].OldValue,
                Is.Null);

            Assert.That(
                events[0].Changes![1].NewValue!.GetString(),
                Is.EqualTo("Public #1"));

            Assert.That(
                events[1].Changes!,
                Is.Null);
        });

        Snapshot.Match(
            events,
            matchOptions => matchOptions
                .IgnoreField("**.Changes")
                .Assert(fieldOption =>
                {
                    Assert.Multiple(() =>
                    {
                        var currentDateTime = DateTime.UtcNow;

                        // id
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Id"),
                            Is.Not.Default);

                        // createdDate
                        Assert.That(
                            fieldOption.Field<DateTime>("[0].CreatedDate"),
                            Is.InRange(startDateTime, currentDateTime));

                        // updatedDate
                        Assert.That(
                            fieldOption.Field<DateTime>("[0].UpdatedDate"),
                            Is.InRange(startDateTime, currentDateTime));

                        // createdDate == updatedDate
                        Assert.That(
                            fieldOption.Field<DateTime>("[0].CreatedDate"),
                            Is.EqualTo(fieldOption.Field<DateTime>("[0].UpdatedDate")));

                        // _eTag
                        Assert.That(
                            fieldOption.Field<Guid>("[0].ETag"),
                            Is.Not.Default);

                        // context.objectId
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Context.ObjectId"),
                            Is.Not.Default);

                        // context.httpTraceIdentifier
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Context.HttpTraceIdentifier"),
                            Is.Not.Default);

                        // context.httpRequestPath
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Context.HttpRequestPath"),
                            Is.Not.Default);

                        // id
                        Assert.That(
                            fieldOption.Field<Guid>("[1].Id"),
                            Is.Not.Default);

                        // createdDate
                        Assert.That(
                            fieldOption.Field<DateTime>("[1].CreatedDate"),
                            Is.InRange(startDateTime, currentDateTime));

                        // updatedDate
                        Assert.That(
                            fieldOption.Field<DateTime>("[1].UpdatedDate"),
                            Is.InRange(startDateTime, currentDateTime));

                        // createdDate == updatedDate
                        Assert.That(
                            fieldOption.Field<DateTime>("[1].CreatedDate"),
                            Is.EqualTo(fieldOption.Field<DateTime>("[1].UpdatedDate")));

                        // _eTag
                        Assert.That(
                            fieldOption.Field<Guid>("[1].ETag"),
                            Is.Not.Default);

                        // context.objectId
                        Assert.That(
                            fieldOption.Field<Guid>("[1].Context.ObjectId"),
                            Is.Not.Default);

                        // context.httpTraceIdentifier
                        Assert.That(
                            fieldOption.Field<Guid>("[1].Context.HttpTraceIdentifier"),
                            Is.Not.Default);

                        // context.httpRequestPath
                        Assert.That(
                            fieldOption.Field<Guid>("[1].Context.HttpRequestPath"),
                            Is.Not.Default);

                        // context.objectId
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Context.ObjectId"),
                            Is.EqualTo(fieldOption.Field<Guid>("[1].Context.ObjectId")));

                        // context.httpTraceIdentifier
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Context.HttpTraceIdentifier"),
                            Is.EqualTo(fieldOption.Field<Guid>("[1].Context.HttpTraceIdentifier")));

                        // context.httpRequestPath
                        Assert.That(
                            fieldOption.Field<Guid>("[0].Context.HttpRequestPath"),
                            Is.EqualTo(fieldOption.Field<Guid>("[1].Context.HttpRequestPath")));
                    });
                }));
    }
}
