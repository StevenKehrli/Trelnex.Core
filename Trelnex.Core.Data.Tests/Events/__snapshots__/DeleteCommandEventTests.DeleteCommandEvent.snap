[
  {
    "SaveAction": "CREATED",
    "RelatedId": "bc4971e4-6dae-45ba-b3ee-43c036e0d957",
    "RelatedTypeName": "item-test",
    "Changes": [
      {
        "PropertyName": "publicId",
        "OldValue": {
          "ValueKind": "Number"
        },
        "NewValue": {
          "ValueKind": "Number"
        }
      },
      {
        "PropertyName": "publicMessage",
        "OldValue": null,
        "NewValue": {
          "ValueKind": "String"
        }
      }
    ],
    "Context": {
      "ObjectId": "587e11f5-ffde-4bb7-8c2c-a6778e14f21c",
      "HttpTraceIdentifier": "10493abd-aedc-402f-9ca7-244407aac87f",
      "HttpRequestPath": "69e2466f-53e8-4040-8f77-f723f89a2546"
    },
    "Id": "98e68844-e860-4e67-b56f-e503e34f5053",
    "PartitionKey": "60525a8e-b084-4a37-b461-d08330760ef2",
    "TypeName": "event",
    "CreatedDate": "2024-10-19T00:32:26.05742Z",
    "UpdatedDate": "2024-10-19T00:32:26.05742Z",
    "DeletedDate": null,
    "IsDeleted": null,
    "ETag": "eb80c99b-54b0-4b83-a9a7-95e7e0e5eb85"
  },
  {
    "SaveAction": "DELETED",
    "RelatedId": "bc4971e4-6dae-45ba-b3ee-43c036e0d957",
    "RelatedTypeName": "item-test",
    "Changes": null,
    "Context": {
      "ObjectId": "587e11f5-ffde-4bb7-8c2c-a6778e14f21c",
      "HttpTraceIdentifier": "10493abd-aedc-402f-9ca7-244407aac87f",
      "HttpRequestPath": "69e2466f-53e8-4040-8f77-f723f89a2546"
    },
    "Id": "e475fa35-beec-4c33-9109-40ab7f666af6",
    "PartitionKey": "60525a8e-b084-4a37-b461-d08330760ef2",
    "TypeName": "event",
    "CreatedDate": "2024-10-19T00:32:26.057544Z",
    "UpdatedDate": "2024-10-19T00:32:26.057544Z",
    "DeletedDate": null,
    "IsDeleted": null,
    "ETag": "8735843a-f134-426b-9fd3-2d040d956b39"
  }
]
